### DISCUSSION TITLE
Build v86 for one program/purpose: TTS

### DISCUSSION THREAD

**Discussion by guest271314** [copy/v86#880] [Q&A] [answered]
How to build this containing a single program: eSpeak NG?

Why: Web Speech API is broken; extra work is needed for `speak()` to work on Firefox; neither Chrome nor Firefox support SSML input, nor capturing audio output of the local speech synthesis engine.

I don't need everything else, just the capability to run the eSpeak NG programm and get `--stdout` which I can parse to create a live `MediaStreamTrack` from in the browser.

Thanks in advance for any help with this.

**Comment by cemalgnlts** [comment]
Yes, I'm wondering if anyone has done this too.

I'm building a linux with buildroot just to run nodejs: [cemalgnlts/now-linux-build](https://github.com/cemalgnlts/now-linux-build)

Instead of recompiling Nodejs, I'm using the prebuilt binary so it takes less time to compile I've shrunk busybox to only support a few commands.

Without Nodejs it's ~4MB but not small enough. I'm sure there are many more files that Nodejs doesn't need.

**Comment by guest271314** [comment]
I would suggest experimenting with QuickJS `qjs` which is 5MB, less than 1mb after `strip` unless you have a compelling need to use `node` executable.

**Reply by cemalgnlts** [reply]
I also looked at it as an alternative but I want Node projects to run directly but for QuickJS also Node APIs should be written which takes a lot of effort.

**Reply by guest271314** [reply]
Why do you need Node.js API's?

**Reply by guest271314** [reply]
Note, it is possible to run `node` _controlled_ from the browser using Native Messaging https://github.com/guest271314/native-messaging-nodejs.

**Reply by cemalgnlts** [reply]
Yes, I've looked at that before. I'm building a project more like WebContainers - Nodebox.

**Comment by guest271314** [comment]
Good luck.

**Reply by cemalgnlts** [reply]
Thanks.

**Reply by guest271314** [reply]
Browsing through your source code it looks like embedding eSpeak NG in buildroot should work.

Did you use Rust at all during the build process?

**Reply by guest271314** [reply]
> The Linux build was not built specifically for this project, so there are unnecessary files.

That's what I am trying to avoid.

**Reply by cemalgnlts** [reply, marked as answer]
You need to remove the addons for Node:

[build.sh](https://github.com/cemalgnlts/now-linux-build/blob/03be06d2e879e3bc4bf4651d97f694eb2636e250/build.sh#L10-L33)
[inittab](https://github.com/cemalgnlts/now-linux-build/blob/03be06d2e879e3bc4bf4651d97f694eb2636e250/buildroot-custom/board/now/rootfs_overlay/etc/inittab#L7-L8)

Then run the `./build.sh` command.

When the process is complete the linux image will be in this folder: `buildroot-2023.02/output/images/rootfs.iso9660`

I am using https://buildroot.org to build the Linux image.

I've been looking for a way to make it even smaller, but more is beyond me.

I started building the project from this repo: https://github.com/humphd/browser-vm.

**Reply by guest271314** [reply]
The more I keep looking at this the more I realize the working solution I have with Native Messaging is probably the best fit for me.

I don't want to create a duplicate copy of a program to carry around for use in the browser when I already have the program installed on the machine.

I can just run `node`, `deno`, `bun`, `qjs,` `tjs`, and `espeak-ng` from the browser - without compiling a duplicate version of those programs to run in the browser.

I was looking at your code here `[scripts/boot.js](https://github.com/cemalgnlts/now-linux-build/blob/master/scripts/boot.js)` and it looks like you are using Node.js to build the program, then you download `node` again, just to use in the browser.

**Reply by cemalgnlts** [reply]
v86 allows you to keep the current state while linux is running. I build the linux image using [CircleCI](https://app.circleci.com/pipelines/github/cemalgnlts)\([.circleci/config.yml](https://github.com/cemalgnlts/now-linux-build/blob/master/.circleci/config.yml)\) (GitHub Actions alternative), then I run v86 in the node environment and save the current state when the linux boot process is finished. In this way, I run it from the state file I saved instead of linux.iso and I don't have to wait for the boot process every time.

_Anyone can run Node on their device, but embedding it in the browser will make it more accessible, interactive documents etc._

**Reply by guest271314** [reply]
> Anyone can run Node on their device, but embedding it in the browser will make it more accessible, interactive documents etc.

I can access the `node` executable I download from the browser, run any program I want using that executable and pipe the stdout from that program or `node` itself to the browser. So I don't need to create a duplicate `node` to embed in the browser.

If the `node` you are embedding in the browser cannot access the local file system or run programs then it is a less than complete `node` embedment.

I am just trying to justify continuing to pursue compiling or porting to WASM and I keep encountering the fact that I will be creating a duplicate `espeak-ng` compilation, when I can already use the `espeak-ng` (or `node`) I have installed or accessible (I never actually install `node`, I just use the nightly `node` executable https://github.com/guest271314/download-node-nightly-executable).
